Test include_vars reserved variable matching

s-hertel opened the pull request:
##### SUMMARY

Adding a test for #84624

##### ISSUE TYPE

- Test Pull Request

s-hertel committed:
Add test case for #84624

s-hertel committed:
revert fix temporarily to validate test

s-hertel committed:
reapply fix

mattclay commented on test/integration/targets/var_reserved/tasks/main.yml:
```diff
@@ -21,3 +21,13 @@
+
+- assert:
+    that:
+      - '"Found variable using reserved name: q" not in test_vm_set_host_variable.stderr'
```
Checking only for the absence of a warning is fragile, since the message could change in the future. To ensure that doesn't go undetected, include a second variable which is reserved and assert that warning is seen.

s-hertel replied:
@mattclay I think it's already tested above. #84624 fixed erroneous warnings, and had no bearing on correct warnings. To make that more clear, I could delete the line fixed in #84624 - it's redundant. What do you think?

If I add a test for a variable named `q` giving a warning in include_vars, it will warn correctly with or without the change in #84624.

If I use any other reserved variable name, it will warn correctly with or without the change in #84624.

I'm mainly trying to add tests to unblock my backports for #84624, so I'm specifically trying to test that bug. Would you consider it less fragile if I just search for "Found variable using reserved name" like the tests for reserved variables above?

s-hertel replied:
Since it's a public method, I did not delete the line modified in #84624 even though it's redundant. As long as it's correct, the display will deduplicate warnings.

mattclay replied:
It is tested above, but not in a way that guarantees the negative assertion is valid. Although the test above looks for most of the warning message, it doesn't include the trailing `: {variable_name}` in the assertion. If that portion of the message changed, then the tests could pass even if the bug returned.

A simple solution is to add a second variable to `set_host_variable.yml` that's reserved, such as `lookup`. Then add a second assertion that verifies the warning is emitted:

```yaml
      - '"Found variable using reserved name: lookup" in test_vm_set_host_variable.stderr'
```

That way if the message ever changes, the test will fail and we'll have an opportunity to update both assertions to maintain the expected test behavior. I'd also include a comment explaining the purpose of the second assert, in case someone decides it can be removed later, not realizing why it was added in the first place.

s-hertel replied:
My point is that such a test will pass even if the bug returns. I'm not sure how to make the negative assertion less fragile.

s-hertel replied:
```suggestion
      - 'canary ~ ": q" not in test_vm_set_host_variable.stderr'
```

s-hertel replied:
Sorry, I didn't mean to mark that as resolved, commiting the suggestion did that I think.

s-hertel committed:
erase line

s-hertel committed:
test working include_vars warning

s-hertel committed:
Update test/integration/targets/var_reserved/tasks/main.yml

mattclay commented on lib/ansible/vars/manager.py:
```diff
@@ -594,7 +594,6 @@ def set_host_variable(self, host, varname, value):
         Sets a value in the vars_cache for a host.
         """
 
-        warn_if_reserved([varname])
```
Why is this line being removed?

s-hertel replied:
This is the line that received the bugfix in https://github.com/ansible/ansible/pull/84624 and removing it demonstrates that this line was the cause of the erroneous q warning, and not solely responsible for the valid warnings for reserved variables. Maybe an integration test is the wrong approach to testing this.

s-hertel replied:
I put it back.

s-hertel committed:
fix indentation

test variable for include_vars gives a warning for a variable named q

s-hertel committed:
Add back warning in set_host_variable

s-hertel committed:
remove fragile test

s-hertel committed:
add test for include_vars specific codepath

check number of warnings for an include_vars resolved variable name that contains the letter q

mattclay commented on test/integration/targets/var_reserved/tasks/main.yml:
```diff
@@ -12,12 +12,14 @@
+        - (item['stderr'] | regex_findall(canary) | length) == 1
     loop: '{{play_out.results}}'
     loop_control:
       label: '{{item.item}}'
```
Here's an alternative approach to testing that will prevent the regression while also protecting against changes to the warning message going undetected:

```yaml
- name: check output for warning
  block:
  - shell: ansible-playbook '{{[ role_path, "tasks", item.file ~ ".yml"] | path_join }}'
    environment:
      ANSIBLE_LOCALHOST_WARNING: 0
      ANSIBLE_FORCE_COLOR: 0
    failed_when: false
    loop:
    - file: play_vars
      name: lipsum
    - file: block_vars
      name: query
    - file: task_vars
      name: query
    - file: task_vars_used
      name: q
    - file: set_fact
      name: lookup
    - file: include_vars
      name: query
    register: play_out

  - name: check they all complain about bad defined var
    assert:
      that:
        - item.stderr == warning_message
    loop: '{{play_out.results}}'
    loop_control:
      label: '{{item.item.file}}'
    vars:
      warning_message: "[WARNING]: Found variable using reserved name: {{ item.item.name }}"
```

s-hertel replied:
added

s-hertel committed:
refactor how a single warning is asserted

include var name for good measure
